### PR TITLE
Better glowing lights

### PR DESCRIPTION
**Pull request by actioninja**
## About The Pull Request

A (sort of) more proper implementation of an emissive layer:
image
Only mobs and items will block glow when on the normal emissive plane atm.

Uses filters for the bulk of the work so the effect itself should have minimum server impact.

Works via rendering an extra overlay on an "emissive" plane that's an unrendered render_target slate. The lighting plane then uses this as a masking render_source via the alpha filter.

Blocking works via a second plane that also is an unrendered slate. Instead this one masks out the emissive plane.
atom/movable has a new `blocks_emissive` var which has three values, false, generic, or unique.
Generic would be for anything that doesn't meaningfully change outline or appearance during standard usage. IE, the vast majority of items and simplemobs. It leverages the vis_overlays subsystem to cache the needed overlays so that there is effectively only one overlay needed per object type and direction.
Unique would be for anything that can have very disparate appearances, such as humans. It works via a second helper atom/movable that gets added to the mob's contents. the blocker exists on the emissive blocker plane, and has the parent set as the render source.  This means it's a 1:1 copy of the parent but on the blocker plane. It's then added to vis_contents which makes it match up perfectly.

## Why It's Good For The Game

Corrects the issue with glowing things rendering above everything else.
accurate per-pixel blocking

## Changelog
:cl:
fix: Glowing overlays no longer render above everything, and mobs and items properly block them
/:cl:

### CODE REVIEW

**Comment by MrPerson** [comment]
I really, really, don't think this is acceptable with mobs only. Consider anything on the floor like the atmos filters. Stuff's going to get on top and make things look jankier than the coolness the feature enables in the first place.

I think there's a path for that, not 100% sure how but I have some ideas I can try out.

**Comment by actioninja** [comment]
> I really, really, don't think this is acceptable with mobs only. Consider anything on the floor like the atmos filters. Stuff's going to get on top and make things look jankier than the coolness the feature enables in the first place.
> 
> I think there's a path for that, not 100% sure how but I have some ideas I can try out.

It's not ideal but it's an improvement.

Consider that the screenshot in the OP would look like this on master:
image

**Comment by RaveRadbury** [comment]
This is noticeable whenever a new command report is created, I'll be glad to see this improvement instead.

**Comment by optimumtact** [comment]
please explain how

@MrPerson this is a lot better than what we currently have?

**Comment by actioninja** [comment]
@optimumtact op updated with tech info
@MrPerson figured out a solution for objects and mobs blocking without much overhead, wooo!

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -43,6 +43,27 @@
...
+	var/blocks_emissive = FALSE
+	///Internal holder for emissive blocker object, do not use directly use blocks_emissive
+	var/atom/movable/emissive_blocker/em_block
+
+
+/atom/movable/Initialize(mapload)
+	. = ..()
+	switch(blocks_emissive)
+		if(EMISSIVE_BLOCK_GENERIC)
+			SSvis_overlays.add_vis_overlay(src, icon, icon_state, EMISSIVE_BLOCKER_LAYER, EMISSIVE_BLOCKER_PLANE)
```
Comment (AnturK) [inline code/game/atoms_movable.dm]: You should probably add some helper to update these.

File: code/modules/lighting/emissive_blocker.dm [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	plane = EMISSIVE_BLOCKER_PLANE
+	layer = EMISSIVE_BLOCKER_LAYER
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	//Why?
+	//render_targets copy the transform of the target as well, but vis_contents also applies the transform
+	//to what's in it. Applying RESET_TRANSFORM here makes vis_contents not apply the transform.
+	//Since only render_target handles transform we don't get any applied transform "stacking"
+	appearance_flags = RESET_TRANSFORM
+
+/atom/movable/emissive_blocker/Initialize(mapload, source)
```
Comment (AnturK) [inline code/modules/lighting/emissive_blocker.dm]: You'll need all the overrides from lighting objects on these. Getting exploded etc. Would be nice without copypaste but might be ugly.

**Review by AnturK — COMMENTED** [review]
Very nice, will need a testmerge.

**Comment by TheVekter** [comment]
Will this fix the thing where papers on computer consoles show up beneath the lighting layer?

**Comment by actioninja** [comment]
yes

File: code/modules/lighting/emissive_blocker.dm (hunk 2 above)
Comment (actioninja) [reply code/modules/lighting/emissive_blocker.dm]: copypaste is cleaner considering they're just stub overrides.

File: code/game/atoms_movable.dm (hunk 1 above)
Comment (actioninja) [reply code/game/atoms_movable.dm]: These weren't really intended to be updated, but sure.

File: code/__DEFINES/layers.dm [hunk 3]
```diff
@@ -78,6 +78,19 @@
 #define MASSIVE_OBJ_LAYER 11
 #define POINT_LAYER 12
 
+
+#define EMISSIVE_BLOCKER_PLANE 12
+#define EMISSIVE_BLOCKER_LAYER 12
+#define EMISSIVE_BLOCKER_RENDER_TARGET "*EMISSIVE_BLOCKER_PLANE"
```
Comment (ninjanomnom) [inline code/__DEFINES/layers.dm]: All planes should probably get a render target by default

File: code/game/atoms_movable.dm [hunk 4]
```diff
@@ -43,6 +43,27 @@
...
+		if(EMISSIVE_BLOCK_UNIQUE)
+			render_target = ref(src)
+			em_block = new(src, render_target)
+			vis_contents += em_block
+
+/atom/movable/Destroy()
+	QDEL_NULL(em_block)
+	. = ..()
```
Comment (ninjanomnom) [inline code/game/atoms_movable.dm]:
```suggestion
	return ..()

```

File: code/game/machinery/computer/_computer.dm [hunk 5]
```diff
@@ -42,12 +42,11 @@
 	. += icon_keyboard
 
 	// This whole block lets screens ignore lighting and be visible even in the darkest room
-	// We can't do this for many things that emit light unfortunately because it layers over things that would be on top of it
 	var/overlay_state = icon_screen
 	if(machine_stat & BROKEN)
 		overlay_state = "[icon_state]_broken"
 	SSvis_overlays.add_vis_overlay(src, icon, overlay_state, layer, plane, dir)
```
Comment (ninjanomnom) [inline code/game/machinery/computer/_computer.dm]: You can remove this one, it was only here as the backdrop to the half opacity one

File: code/modules/lighting/emissive_blocker.dm [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+/atom/movable/emissive_blocker/blob_act()
+	return
+
+/atom/movable/emissive_blocker/onTransitZ()
+	return
+
+//Prevents people from moving these after creation, because they shouldn't be.
+/atom/movable/emissive_blocker/forceMove(atom/destination, no_tp=FALSE, harderforce = FALSE)
+	if(harderforce)
+		. = ..()
```
Comment (ninjanomnom) [inline code/modules/lighting/emissive_blocker.dm]:
```suggestion
		return ..()
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/__DEFINES/layers.dm (hunk 3 above)
Comment (actioninja) [reply code/__DEFINES/layers.dm]: I don't know if just setting render_target incurs any kind of performance penalty so I'll add the defines but not set them in the file without a confirmation.

Comment (ninjanomnom) [reply code/__DEFINES/layers.dm]: I meant that in New you should assign them a render target based on the plane number or something

File: code/__DEFINES/layers.dm [hunk 7]
```diff
@@ -4,11 +4,16 @@
...
+#define PLANE_SPACE_RENDER_TARGET "PLANE_SPACE"
 #define PLANE_SPACE_PARALLAX -90
+#define PLANE_SPACE_PARALLAX_RENDER_TARGET "PLANE_SPACE_PARALLAX"
 
 #define FLOOR_PLANE -2
+#define FLOOR_PLANE_RENDER_TARGET "FLOOR_PLANE"
 #define GAME_PLANE -1
+#define GAME_PLANE_RENDER_TARGET "GAME_PLANE"
 #define BLACKNESS_PLANE 0 //To keep from conflicts with SEE_BLACKNESS internals
+#define BLACKNESS_PLANE_RENDER_TARGET "BLACKNESS_PLANE"
```
Comment (ninjanomnom) [inline code/__DEFINES/layers.dm]: Yeah not this, I meant making it automatic in the planemaster New based on the plane number. Something like
```
#define PLANE_RENDER_TARGET(plane) "*planemaster_[plane]"

/obj/screen/plane_master/New()
  render_target = PLANE_RENDER_TARGET(plane)

/something/else/someproc()
  render_source = PLANE_RENDER_TARGET(BLACKNESS_PLANE)
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/__DEFINES/layers.dm (hunk 7 above)
Comment (actioninja) [reply code/__DEFINES/layers.dm]: I started implementing this and imo that some plane masters are rendering and some aren't makes this overly fiddly to work with. You have to know ahead of time what type a plane is before you use it anywhere no matter how I can think of implementing this, which is a lot less usable than the current implementation.

File: code/__DEFINES/layers.dm (hunk 3 above)
Comment (Dennok) [reply code/__DEFINES/layers.dm]: How to work `*` here

File: code/__DEFINES/layers.dm (hunk 7 above)
Comment (ninjanomnom) [reply code/__DEFINES/layers.dm]: Ah right you need the asterisk or not in different cases. Alright this is fine on an individual basis I guess.

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**